auth-jwt-service/src/main/java/io/stargate/auth/jwt/AuthJwtService.java

```diff
@@ -69,6 +79,148 @@ public StoredCredentials validateToken(String token) throws UnauthorizedExceptio
     return storedCredentials;
   }
 
+  /**
+   * Using the provided JWT and the claims it contains will perform pre-authorization where
+   * possible, executes the query provided, and then authorizes the response of the query.
+   *
+   * @param action A {@link QueryBuilder} object to be executed and authorized against a JWT.
+   * @param token The authenticated JWT token to use for authorization.
+   * @param primaryKeyValues A list of primary key values that will be used in the query and should
+   *     be authorized against the JWT.
+   * @param tableMetadata The {@link Table} that will be queried against.
+   * @return On success will return the result of the query and otherwise will return an exception
+   *     relating to the failure to authorize.
+   * @throws Exception An exception relating to the failure to authorize.
+   */
+  @Override
+  public ResultSet executeDataReadWithAuthorization(
+      Callable<ResultSet> action, String token, List<String> primaryKeyValues, Table tableMetadata)
+      throws Exception {
+    JSONObject stargateClaims = extractClaimsFromJWT(token);
+
+    preCheckDataReadWrite(stargateClaims, primaryKeyValues, tableMetadata);
+
+    ResultSet result = action.call();
+
+    if (result == null) {
+      return null;
+    }
+
+    postCheckDataRead(stargateClaims, result);
+
+    return result;
+  }
+
+  /**
+   * TODO
+   *
+   * @param action
+   * @param token
+   * @param primaryKeyValues
+   * @param tableMetadata
+   * @return
+   * @throws Exception
+   */
+  @Override
+  public ResultSet executeDataWriteWithAuthorization(
+      Callable<ResultSet> action, String token, List<String> primaryKeyValues, Table tableMetadata)
+      throws Exception {
+    JSONObject stargateClaims = extractClaimsFromJWT(token);
+
+    preCheckDataReadWrite(stargateClaims, primaryKeyValues, tableMetadata);
+
+    // Just return the result. No value in doing a post check since we can't roll back anyway.
+    return action.call();
+  }
+
+  @Override
+  public ResultSet executeSchemaReadWithAuthorization(
+      Callable<ResultSet> action, String token, String keyspace, String table) throws Exception {
```

Comment:
Since we verify access to the specified table, would it make sense to pass it as a parameter to the `action` to protect against mistakes of using different tables inside the action?

Reply:
Sure, I can look into that in the following PR for schema related actions.